rest/admin_api_test.go

```diff
@@ -801,59 +801,51 @@ func TestSessionTtlGreaterThan30Days(t *testing.T) {
-		t.Skip("skipping test in short mode")
-	}
-
 	rt := NewRestTester(t, nil)
 	defer rt.Close()
 
 	a := auth.NewAuthenticator(rt.Bucket(), nil)
-	user, err := a.GetUser("")
-	assert.NoError(t, err)
-	user.SetDisabled(true)
-	err = a.Save(user)
-	assert.NoError(t, err)
-
-	user, err = a.GetUser("")
-	assert.NoError(t, err)
-	goassert.True(t, user.Disabled())
-
-	response := rt.SendRequest("PUT", "/db/doc", `{"hi": "there"}`)
-	assertStatus(t, response, 401)
-
-	user, err = a.NewUser("pupshaw", "letmein", channels.SetOf(t, "*"))
+	user, err := a.NewUser("pupshaw", "letmein", channels.SetOf(t, "*"))
 	a.Save(user)
 
-	assertStatus(t, rt.SendAdminRequest("GET", "/db/_session", ""), 200)
+	if err != nil {
+		log.Printf(err.Error())
+	}
 
-	response = rt.SendAdminRequest("POST", "/db/_session", `{"name":"pupshaw", "ttl":1}`)
-	assertStatus(t, response, 200)
+	// More than 10% of expiry (24h)
+	fakeSession := auth.LoginSession{
+		ID:         base.GenerateRandomSecret(),
+		Username:   "user1",
+		Expiration: time.Now().Add(time.Hour * 4),
+		Ttl:        time.Hour * 24,
+	}
 
-	var body db.Body
-	base.JSONUnmarshal(response.Body.Bytes(), &body)
-	sessionId := body["session_id"].(string)
-	sessionExpiration := body["expires"].(string)
-	assert.NotEmpty(t, sessionId)
-	assert.NotEmpty(t, sessionExpiration)
-	assert.True(t, body["cookie_name"].(string) == "SyncGatewaySession")
+	rt.Bucket().Set(docIDForSession(fakeSession.ID), 0, fakeSession)
 
 	reqHeaders := map[string]string{
-		"Cookie": "SyncGatewaySession=" + body["session_id"].(string),
+		"Cookie": "SyncGatewaySession=" + fakeSession.ID,
 	}
-	response = rt.SendRequestWithHeaders("PUT", "/db/doc1", `{"hi": "there"}`, reqHeaders)
-	assertStatus(t, response, 201)
+	response := rt.SendRequestWithHeaders("PUT", "/db/doc1", `{"hi": "there"}`, reqHeaders)
+	log.Printf("PUT Request: Set-Cookie: %v", response.Header().Get("Set-Cookie"))
+	assertStatus(t, response, http.StatusCreated)
+	goassert.True(t, response.Header().Get("Set-Cookie") != "")
 
+	response = rt.SendRequestWithHeaders("GET", "/db/doc1", "", reqHeaders)
+	log.Printf("GET Request: Set-Cookie: %v", response.Header().Get("Set-Cookie"))
+	assertStatus(t, response, http.StatusOK)
 	goassert.True(t, response.Header().Get("Set-Cookie") == "")
 
-	//Sleep for 150ms seconds, this will ensure 10% of the 1 second session ttl has elapsed and
-	//should cause a new Cookie to be sent by the server with the same session ID and an extended expiration date
-	time.Sleep(150 * time.Millisecond)
-	response = rt.SendRequestWithHeaders("PUT", "/db/doc2", `{"hi": "there"}`, reqHeaders)
-	assertStatus(t, response, 201)
+	fakeSession.Expiration = time.Now().Add(time.Hour * 21)
+	rt.Bucket().Set(docIDForSession(fakeSession.ID), 0, fakeSession)
 
+	response = rt.SendRequestWithHeaders("GET", "/db/doc1", "", reqHeaders)
+	log.Printf("GET Request: Set-Cookie: %v", response.Header().Get("Set-Cookie"))
 	goassert.True(t, response.Header().Get("Set-Cookie") != "")
+
```

Comment:
This last request is checking the session is extended again, but I'm not sure how it's different to the first request being made on line 828.

Would it be worth testing an expired session is not allowed to make requests by setting the expiry time to be in the past?

Reply:
I tried firing the GET request with an expired session (an hour past the current time, now()). But it gives OK status code (response code: 200). It looks strange to me. I'm trying to drill down further to understand the root cause of this.

Reply:
OK sure. Like you said in Slack, it could be that we're only relying on the document being removed via the doc TTL to actually expire the session, and actually it seems that is the case:


I'll loop @adamcfraser in on this conversation, as perhaps an additional expiry timestamp check would make sense to make in SG too.
In particular, I'm thinking of the scenario where clocks differ between SG and CB Server nodes, which is something we've seen before.

Reply:
/cc @adamcfraser 

Like you said in Slack, it looks like we're relying on Couchbase Server to expire the document based on the TTL set:

https://github.com/couchbase/sync_gateway/blob/96e0ae5e56003e3a8b3f89895a763aa123b4156a/auth/session.go#L46

I'm wondering if there's value in adding a check for the session expiry timestamp in SG when we try to use a session that isn't expired from Couchbase Server, but still exists with a timestamp in the past.

In particular, I'm thinking of a couple of scenarios:

1. CB and SG clocks are not synchronized, which might mean that we have a non-TTL'd session doc in CB Server, but with an expiry timestamp that is behind SG's clock.
Given SG is setting that initial timestamp via `time.Now().Add(ttl)`, I think we should be using SG's clock to be the source of truth for session expiry, and only rely on TTL'd docs for convenience/cleanup.
2. A bug in CB Server where docs aren't being removed with the TTL. This is unlikely, but being defensive here won't hurt.

Reply:
Code changes are committed as per what we've decided this morning.